### ISSUE TITLE
A new engine with Intel MKL for Theano

### ISSUE THREAD

**Issue by ciyongch** [Theano/Theano#5580]
We are from Intel corporation Ltd., and working on Theano optimization for CPU especially Intel® Xeon® and Intel® Xeon Phi™ processor.

To leverage the performance improvements for all Theano user, we plan to merge our optimization into Theano’s master branch 
and would like to hear your suggestions in advance. 
 
Please check the following details:

# 1 Motivation
Recently, Intel released a new version of MKL® library which provides the high performance primitives to accelerate deep neural network (dnn) on Xeon/Xeon Phi processors.
We have created a new engine under theano/theano/sandbox/mkl as same as cuda for MKL-accelerated Ops. Now, we are doing further verification and performance test for upstreaming in March 2017. 

So we would like to hear from you for this new engine, any comments or suggestion for the PR are highly appreciated.

# 2	Current Status/Perf
We added some new Theano Ops implementation with Intel MKL, including convolution (support bias and group conv), pooling (max pooling and average pooling), relu, normalization(batch norm and local response norm), elemwise sum Ops. 
Also, we added several graph optimizers to automatically replace the original Ops for better performance on Xeon/Xeon Phi processors.
   The portal to the optimized Theano is [here](https://github.com/intel/theano)
    
Currently, we have finished the first version of these Ops. Compared Stock Theano by running [convnet-benchmark](https://github.com/soumith/convnet-benchmarks) on Xeon processor (Dual E5-2699 v4 @ 2.20GHz), the new version shows strongly performance improvement as listed in below table:

| FWD: sec/batch| Stock Theano/Xeon | Optimized Theano/Xeon | speed up |
|---------------|-------------------|-------------------|-----------------------|
| AlexNet       | 1.045             | 0.077             | 13.6                  |
| GoogLeNet     | 2.228             | 0.655             | 3.4                   |
| VGG           | 5.089             | 0.836             | 6.1                   |
| OverFeat      | 6.105             | 0.265             | 23.0                  |

--

| FWD+BWD: sec/batch: sec/batch| Stock Theano/Xeon | Optimized Theano/Xeon | speed up |
|---------------|-------------------|-------------------|-----------------------|
| AlexNet       | 2.333             | 0.236             | 9.9                   |
| GoogLeNet     | 5.866             | 1.535             | 3.8                   |
| VGG           | 12.783            | 2.721             | 4.7                   |
| OverFeat      | 13.202            | 0.830             | 15.9                  |

# 3 Details
Here's the functionality/modules added into current Theano's framework:
*	A new folder "mkl" under "theano/sandbox" for MKL engine, and "mkl/tests" for all the relevant unit tests
*	A group of new config option for MKL: theano.config.mkl
	- Theano.config.mkl.lib: 
		- mkl: default value.
		- mkl-dnn: (another version of MKL which is an open source library dedicated for neural networks, plan to support it in future)
	- Theano.config.mkl.nn.enabled:
		- auto: default value. Automatically detect and use MKL dnn primitives if it is available. If MKL is unavailable, raise no error.
		- True: require the use of MKL. If MKL is unavailable, raise an error.
		- False: do not use MKL or check if it is available

* New Theano Ops implemented with Intel MKL

|Modules |Files | Op Names| Descriptions |
|--------|------|---------|--------------|
|Convolution |mkl_conv.py |Conv2D| 2D Convolution with MKL Op|
|		| |ConvGradWeight|	Convolution gradient w.r.t weight with MKL Op|
|		| |ConvGradImage|	Convolution gradient w.r.t image(data) with MKL Op|
|		| |AbstractConvGroup|	Abstract group convolution, it's a place holder, will be replaced by conv MKL Op|
|		| |AbstractConvGroupGrad|	Abstract group convolution gradient, it's a place holder, will be replaced by conv gradient MKL Op|
|Pool|mkl_pool.py|	Pool|	Pooling with MKL Op|
|		| |PoolGrad|	Pooling gradient with MKL Op|
|Relu	|mkl_relu.py|	AbstractRelu|	Abstract relu, it's a place holder, will be replaced by Relu MKL Op|
|		| |AbstractReluGrad|	Abstract relu gradient, it's a place holder, will be replaced by Relu gradient MKL Op|
|		| |Relu|	Relu with MKL Op|
|		| |ReluGrad|	Relu gradient with MKL Op|
|LocalResponseNorm|	mkl_rln.py|	AbstractLocalResponseNorm|	Abstract lrn, it's a place holder, will be replaced by LRN MKL Op.|
|		| |AbstractLocalResponseNormGrad|	Abstract lrn gradient, it's a place holder, will be replaced by LRN gradient MKL Op.|
|		| |LocalResponseNorm|	LRN with MKL Op|
|		| |LocalResponseNormGrad|	LRN gradient with MKL Op|
|batch Norm|	mkl_bn.py|	AbstractBatchNormalization|	Abstract bn, it's a place holder, will be replaced by BN MKL Op.|
|		| |AbstractBatchNormalizationGrad|	Abstract bn gradient, it's a place holder, will be replaced by BN gradient MKL Op|
|		| |BatchNormalization|	BN with MKL op|
|		| |BatchNormalizationGrad|	BN gradient with MKL Op|
|ElemwiseSum|	mkl_elemwise.py|	ElemwiseSum	elemwise sum with MKL Op|
|		| |ElemwiseSumGrad|	elemwise sum gradient with MKL Op|

* New graph optimizers for Intel MKL Ops

|Modules|	Files|	Op/Function Names|	Descriptions|
|-------|--------|-------------------|--------------|
|local Optimizer|	opt.py|	local_Conv2D_mkl|	local group optimizer for conv2d|
|		| |local_ConvGradInputs_mkl|	local group optimizer for conv gradient w.r.t inputs|
|		| |local_ConvGradWeights_mkl|	local group optimizer for conv gradient w.r.t weights|
|		| |local_pool_mkl|	local optimizer for pool|
|		| |local_poolGrad_mkl|	local optimizer for pool gradient|
|		| |local_relu_mkl|	local optimizer for relu|
|		| |local_reluGrad_mkl|	local optimizer for relu gradient|
|		| |local_lrn_mkl	local| optimizer for lrn|
|		| |local_lrnGrad_mkl|	local optimizer for lrn gradient|
|		| |local_bn_mkl|	local optimizer for bn|
|		| |local_bnGrad_mkl|	local optimizer for bn gradient|
|Global Optimizer|	|	ReplaceElemwise|	global optimizer for elemwise sum|
|		| |ReplaceConvBias|	global optimizer for convolution with bias|
|		| |CutMKLDataConversionChain|	global optimizer for cuting useless MKL data conversion chain|

**Comment by SvenMeyer** [comment]
The GoogLeNet time or speedup does not look right to me:

FWD:sec/batch | StockTheano/Xeon | OptimizedTheano/Xeon | speedup
AlexNet 	| 1.045 	| 0.077 	| 13.6
GoogLeNet 	|  2.228 	| 0.065 	| 3.4

**Comment by nouiz** [comment]
Thanks for the heads up about this.

A few comments. I didn't do an intensive review, just a very quick look.

The change about _MM_SET_FLUSH_ZERO_MODE should be user controlable ideally
via a Theano flags. The could could be read at the import of theano and
call a function in the cutils.py file instead of the lazy_linker stuff.

The change to dnn.enable with just a look at the doc string seem a good
user interface.

For the abstrct ops, when Theano already have one, like for the convolution
and BN (very recently added), you should reuse Theano abstract op and not
create a new one. Why did you needed a new abstract conv op? Is it because
it support a feature the current abstract op don't support like group
convolution? If so, we should discuss that with @lamblin, maybe it would be
better to add a new parameter to the current abstract conv instead of
making a new one.

**Comment by pcs-theano** [comment]
@SvenMeyer It's a typo for the timing data, should be 0.655 but not 0.065, I've updated the table. Thanks for pointing it out.


@nouiz 
Thanks for your valuable comments.

We will change _MM_SET_FLUSH_ZERO_MODE to be controllable via Theano flags such as
"theano.config.enableFTZ=True/False, theano.config.enableDAZ=True/False", and move
the code from lazy_linker_c.c to cutils.py as you suggested.

Regarding dnn.enable, we did use this config to extend for MKL before. But the current
one might not be enough to cover all the MKL usages in future, besides this will change
 the current config's behavior. So we use a new set config options 
 "theano.config.mkl.lib" and "theano.config.mkl.nn.enabled" for MKL settings.
Actually what you saw on the master branch yesterday was not our up-to-date code base. 
Anyway, I've merged our latest code into the master branch, please take a look when you
have time.

For the abstractConvGroup op, just as you mentioned the current abstract conv 
can't support this feature, if it's acceptable to add a new parameter "group" to the
current one, it's great and we can do this.

For the abstractRelu op, the current relu function was built up with several basic (elemwise) ops which was hard for find it during graph optimization phase, so we added abstractRelu
just as a place holder which will be called by nnet.relu to be replaced with MKL relu primitives during
compiling.
As for abstractBN op, we're considering to reuse the latest one, and update the code later.

For the abstractLRN op, we added this since there's no LRN op in current Theano, we can provide this op since we already have it although a few people use it.

BTW, except convnet-benchmarks, are there any other workloads or benchmarks preferred to be verified?

**Comment by nouiz** [comment]
for abstractConvGroup. I had a quick discussion with @lamblin and he think a new op is the way to go. So for now, don't change that.

For the relu, the abstract op make sence I think. The question that I didn't looked at in the code is when we need to do the substitution. We should do it when mkl isn't there at a timing that won't make us loose other Theano optimization about it.

For LRN, keep the abstract op. It will be useful later.

For the benchmark, convnet-benchmark is good for convnet. We have a daily speed test buildbot that use the deep learning tutorial example for speed tests. Not sure it they are useful for your case. If you are interrested, it is in the file tests.py but in the function speed in that file.
There is also the benchmark we did in the last Theano article: https://github.com/caglar/rnn_benchmarks, and maybe https://github.com/yaoli/arctic-capgen-vid

**Comment by f0k** [comment]
+1 for an abstract relu in Theano (or just a relu op with a default implementation that can be replaced), see #2698. Consider a `leakiness` parameter as well, to also support LReLU, PReLU and RReLU.

**Comment by pcs-theano** [comment]
@nouiz 
For the abstractRelu, our current design is only to do the substitution when MKL is available for use, otherwise, will keep the original relu function which implemented by elemwise and go with current Theano optimizations.
So we will keep the current abstractConvGroup, abstractRelu, abstractLRN, and reuse the latest BN op.
Let me try with the benchmarks you mentioned above.


@f0k The optimal relu op supports slope parameter, so it's ok to support the mode you mentioned.
